Bug Report: vtb e2e test for stats verification is flaky

### Overview of the Issue

VTB E2E test that does stats verification is some times flaky, I _think_ due to `servenv.RunDefault` goroutine racing with stats writing.

### Reproduction Steps

This does not appear to be happening in the vitessio/vitess origin, but is happening in our (PlanetScale's) private fork's CI, even though it does not happen locally on my laptop with that fork.

Not 100% sure what is causing it to fail in the private fork CI, but not locally nor in the upstream CI.

```
=== RUN   TestTabletInitialBackup
I0927 17:53:17.078700   31163 backup_only_test.go:215] starting backup tablet 2023-09-27 17:53:17.078698562 +0000 UTC m=+38.995604489
I0927 17:53:17.078740   31163 cluster_process.go:1103] Starting vtbackup
I0927 17:53:17.078746   31163 cluster_process.go:1131] Attempting to reserve port: 16018
I0927 17:53:17.078793   31163 cluster_process.go:1139] Port 16018 is available, reserving..
I0927 17:53:17.139357   31163 vtbackup_process.go:87] Running vtbackup with args: vtbackup --topo_implementation etcd2 --topo_global_server_address localhost:16002 --topo_global_root /vitess/global --log_dir /tmp/vt_2451224871/vtroot_16001/tmp_16003 --mysql_port 16018 --init_db_sql_file /tmp/vt_2451224871/vtroot_16001/tmp_16003/init_db_with_passwords.sql --init_keyspace ks --init_shard 0 --backup_storage_implementation file --file_backup_storage_root /tmp/vt_2451224871/vtroot_16001/tmp/backupstorage --initial_backup --allow_first_backup --db-credentials-file /tmp/vt_2451224871/vtroot_16001/tmp_16003/db_credentials.json --mysql_socket /tmp/vtbackup_test_mysql.sock4005173465 --stats_backend opentsdb --opentsdb_uri file:///tmp/TestTabletInitialBackup4259381491/001/opentsdb.TestTabletInitialBackup.txt
W0927 17:53:17.179987   33149 log.go:39] Failed to read in config : Config File "vtconfig" Not Found in "[/opt/actions-runner/_work/vitess-private/vitess-private/go/test/endtoend/backup/vtbackup]". This is optional, and can be ignored if you are not using config files. For a detailed explanation, see https://github.com/vitessio/vitess/blob/main/doc/viper/viper.md#config-files.
I0927 17:53:34.034417   31163 vtctl_process.go:92] Executing vtctlclient with arguments vtctl --log_dir /tmp/vt_2451224871/vtroot_16001/tmp_16003 --topo_implementation etcd2 --topo_global_server_address localhost:16002 --topo_global_root /vitess/global --backup_storage_implementation file --file_backup_storage_root /tmp/vt_2451224871/vtroot_16001/tmp/backupstorage ListBackups ks/0
    backup_only_test.go:4[72](https://github.com/planetscale/vitess-private/actions/runs/6327179880/job/17189903826#step:8:73): 
        	Error Trace:	/opt/actions-runner/_work/vitess-private/vitess-private/go/test/endtoend/backup/vtbackup/backup_only_test.go:472
        	            				/opt/actions-runner/_work/vitess-private/vitess-private/go/test/endtoend/backup/vtbackup/backup_only_test.go:67
        	Error:      	Not equal: 
        	            	expected: 0
        	            	actual  : 1
        	Test:       	TestTabletInitialBackup
--- FAIL: TestTabletInitialBackup (47.08s)
```

### Binary Version

```sh
Version: 18.0.0-SNAPSHOT
```


### Operating System and Environment details

```sh
GitHub Actions.
```


### Log Fragments

```sh
=== RUN   TestTabletInitialBackup
I0927 17:53:17.078700   31163 backup_only_test.go:215] starting backup tablet 2023-09-27 17:53:17.078698562 +0000 UTC m=+38.995604489
I0927 17:53:17.078740   31163 cluster_process.go:1103] Starting vtbackup
I0927 17:53:17.078746   31163 cluster_process.go:1131] Attempting to reserve port: 16018
I0927 17:53:17.078793   31163 cluster_process.go:1139] Port 16018 is available, reserving..
I0927 17:53:17.139357   31163 vtbackup_process.go:87] Running vtbackup with args: vtbackup --topo_implementation etcd2 --topo_global_server_address localhost:16002 --topo_global_root /vitess/global --log_dir /tmp/vt_2451224871/vtroot_16001/tmp_16003 --mysql_port 16018 --init_db_sql_file /tmp/vt_2451224871/vtroot_16001/tmp_16003/init_db_with_passwords.sql --init_keyspace ks --init_shard 0 --backup_storage_implementation file --file_backup_storage_root /tmp/vt_2451224871/vtroot_16001/tmp/backupstorage --initial_backup --allow_first_backup --db-credentials-file /tmp/vt_2451224871/vtroot_16001/tmp_16003/db_credentials.json --mysql_socket /tmp/vtbackup_test_mysql.sock4005173465 --stats_backend opentsdb --opentsdb_uri file:///tmp/TestTabletInitialBackup4259381491/001/opentsdb.TestTabletInitialBackup.txt
W0927 17:53:17.179987   33149 log.go:39] Failed to read in config : Config File "vtconfig" Not Found in "[/opt/actions-runner/_work/vitess-private/vitess-private/go/test/endtoend/backup/vtbackup]". This is optional, and can be ignored if you are not using config files. For a detailed explanation, see https://github.com/vitessio/vitess/blob/main/doc/viper/viper.md#config-files.
I0927 17:53:34.034417   31163 vtctl_process.go:92] Executing vtctlclient with arguments vtctl --log_dir /tmp/vt_2451224871/vtroot_16001/tmp_16003 --topo_implementation etcd2 --topo_global_server_address localhost:16002 --topo_global_root /vitess/global --backup_storage_implementation file --file_backup_storage_root /tmp/vt_2451224871/vtroot_16001/tmp/backupstorage ListBackups ks/0
    backup_only_test.go:472: 
        	Error Trace:	/opt/actions-runner/_work/vitess-private/vitess-private/go/test/endtoend/backup/vtbackup/backup_only_test.go:472
        	            				/opt/actions-runner/_work/vitess-private/vitess-private/go/test/endtoend/backup/vtbackup/backup_only_test.go:67
        	Error:      	Not equal: 
        	            	expected: 0
        	            	actual  : 1
        	Test:       	TestTabletInitialBackup
--- FAIL: TestTabletInitialBackup (47.08s)
```
```

--- a/go/cmd/vtbackup/cli/vtbackup.go
+++ b/go/cmd/vtbackup/cli/vtbackup.go
@@ -229,14 +229,12 @@ func run(_ *cobra.Command, args []string) error {
 		<-ctx.Done()
 	}()
 
-	// Some plugins use OnRun to initialize. Use a channel to signal to
-	// ourselves that these plugins have been initialized.
-	runCh := make(chan struct{})
-	servenv.OnRun(func() {
-		close(runCh)
-	})
 	go servenv.RunDefault()
-	<-runCh
+	// Some stats plugins use OnRun to initialize. Wait for them to finish
+	// initializing before continuing, so we don't lose any stats.
+	if err := stats.AwaitBackend(ctx); err != nil {
+		return fmt.Errorf("failed to await stats backend: %w", err)
+	}
 
 	if detachedMode {
 		// this method will call os.Exit and kill this process

--- a/go/cmd/vtbackup/cli/vtbackup.go
+++ b/go/cmd/vtbackup/cli/vtbackup.go
@@ -229,7 +229,14 @@ func run(_ *cobra.Command, args []string) error {
 		<-ctx.Done()
 	}()
 
+	// Some plugins use OnRun to initialize. Use a channel to signal to
+	// ourselves that these plugins have been initialized.
+	runCh := make(chan struct{})
+	servenv.OnRun(func() {
+		close(runCh)
+	})
 	go servenv.RunDefault()
+	<-runCh
 
 	if detachedMode {
 		// this method will call os.Exit and kill this process

--- a/go/stats/export.go
+++ b/go/stats/export.go
@@ -29,6 +29,7 @@ package stats
 
 import (
 	"bytes"
+	"context"
 	"expvar"
 	"fmt"
 	"strconv"
@@ -45,6 +46,7 @@ var (
 	emitStats         bool
 	statsEmitPeriod   = 60 * time.Second
 	statsBackend      string
+	statsBackendInit  = make(chan struct{})
 	combineDimensions string
 	dropVariables     string
 )
@@ -209,6 +211,18 @@ var pushBackends = make(map[string]PushBackend)
 var pushBackendsLock sync.Mutex
 var once sync.Once
 
+func AwaitBackend(ctx context.Context) error {
+	if statsBackend == "" {
+		return nil
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-statsBackendInit:
+		return nil
+	}
+}
+
 // RegisterPushBackend allows modules to register PushBackend implementations.
 // Should be called on init().
 func RegisterPushBackend(name string, backend PushBackend) {
@@ -218,6 +232,9 @@ func RegisterPushBackend(name string, backend PushBackend) {
 		log.Fatalf("PushBackend %s already exists; can't register the same name multiple times", name)
 	}
 	pushBackends[name] = backend
+	if name == statsBackend {
+		close(statsBackendInit)
+	}
 	if emitStats {
 		// Start a single goroutine to emit stats periodically
 		once.Do(func() {

